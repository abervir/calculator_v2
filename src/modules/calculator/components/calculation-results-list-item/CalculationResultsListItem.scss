.controls-column {
  width: 100px;
}
