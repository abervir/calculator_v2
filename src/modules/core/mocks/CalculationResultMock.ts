import { OperationIcons } from '@modules/calculator/enums';
import { CalculationResult } from '@modules/calculator/models';

export const CalculationResultMock: CalculationResult = {
  id: 1,
  result: 1,
  operands: [0, 1],
  operation: {
    id: 'ID',
    title: 'Operation1',
    icon: OperationIcons.PLUS,
    sign: '+',
    operation: (a: number, b: number) => a + b,
  },
};
