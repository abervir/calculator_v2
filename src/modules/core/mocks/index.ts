export * from './CalculationResultMock';
export * from './EventMock';
