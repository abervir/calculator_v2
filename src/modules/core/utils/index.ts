export * from './input.utils';
export * from './shared.utils';
export * from './math.utils';
