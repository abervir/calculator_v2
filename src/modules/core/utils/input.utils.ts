import { KeyboardEvent } from 'react';
import { ALLOWED_INPUT_CHARS } from '../constants';

export class InputUtils {
  static isInputNaN(event: KeyboardEvent<HTMLInputElement>): boolean {
    return (
      isNaN(Number(event.key)) &&
      !ALLOWED_INPUT_CHARS.includes(event.key) &&
      !event.ctrlKey
    );
  }

  static onKeyPressChecker(event: KeyboardEvent<HTMLInputElement>): void {
    if (InputUtils.isInputNaN(event)) {
      event.preventDefault();
    }
  }
}
