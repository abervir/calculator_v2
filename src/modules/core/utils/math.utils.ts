export class MathUtils {
  static getHighestPrimeNumberBetweenTwoNumbers(
    minValue: number,
    maxValue: number
  ): number | null {
    let primeNumber = null;

    for (var i = minValue; i <= maxValue; i++) {
      if (MathUtils.isPrime(i)) {
        primeNumber = i;
      }
    }

    return primeNumber;
  }

  static isPrime(value: number): boolean {
    if (value < 2) {
      return false;
    }

    for (let i = 2; i < value; i++) {
      if (value % i === 0) {
        return false;
      }
    }

    return true;
  }

  static remainderOfDivision(a: number, b: number): number {
    return a % b;
  }

  static divide(a: number, b: number): number {
    return a / b;
  }

  static sum(a: number, b: number): number {
    return a + b;
  }
}
