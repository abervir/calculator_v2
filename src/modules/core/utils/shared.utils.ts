export class SharedUtils {
  static copyTextToClipboard(value: string): void {
    navigator.clipboard.writeText(value);
  }
}
