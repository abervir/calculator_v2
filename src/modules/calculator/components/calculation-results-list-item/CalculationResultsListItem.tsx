import React from 'react';

import { Button } from '@modules/shared/components';
import { SharedUtils } from '@modules/core/utils';
import { CalculationResult } from '@modules/calculator/models';

import './CalculationResultsListItem.scss';

interface Props {
  item: CalculationResult;
}

const _CalculationResultsListItem = (props: Props) => {
  const {
    item: { operands, operation, result, id },
  } = props;

  const resultValue = result ?? 'N/A';
  const operationStringRepresentation = `${operands[0]} ${operation.sign} ${operands[1]} = ${resultValue}`;

  return (
    <>
      <td className='calculation-result-title'>{operation.title} operation</td>
      <td className='calculation-result-representation'>
        {operationStringRepresentation}
      </td>
      <td>{new Date(id).toUTCString()}</td>
      <td className='controls-column'>
        <Button
          title='Copy result!'
          onClick={() => SharedUtils.copyTextToClipboard(result.toString())}
        />
      </td>
    </>
  );
};

export const CalculationResultsListItem = React.memo(
  _CalculationResultsListItem
);
