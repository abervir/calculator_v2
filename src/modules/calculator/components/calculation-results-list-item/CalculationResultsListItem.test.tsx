import { shallow } from 'enzyme';

import { CalculationResultsListItem } from './CalculationResultsListItem';
import { SharedUtils } from '@modules/core/utils';
import { CalculationResultMock } from '@modules/core/mocks';

describe('CalculationResultsListItem', () => {
  it(`should render component`, () => {
    expect(
      shallow(<CalculationResultsListItem item={CalculationResultMock} />)
    ).toMatchSnapshot();
  });

  it(`should invoke copyTextToClipboard after button was pressed`, () => {
    const spy = jest
      .spyOn(SharedUtils, 'copyTextToClipboard')
      .mockImplementation();
    const wrapper = shallow(
      <CalculationResultsListItem item={CalculationResultMock} />
    );
    wrapper.find('Button').simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  it(`should display operation title`, () => {
    const title = `${CalculationResultMock.operation.title} operation`;
    const wrapper = shallow(
      <CalculationResultsListItem item={CalculationResultMock} />
    );
    const titleText = wrapper.find('.calculation-result-title').text();

    expect(titleText).toBe(title);
  });

  [
    {
      result: 1,
    },
    {
      result: null,
    },
  ].forEach(({ result }) => {
    it(`should display ${result} in result represenatation column`, () => {
      const { operands, operation } = CalculationResultMock;
      CalculationResultMock.result = result as number;
      const resultValue = result || 'N/A';
      const operationStringRepresentationExepected = `${operands[0]} ${operation.sign} ${operands[1]} = ${resultValue}`;
      const wrapper = shallow(
        <CalculationResultsListItem item={CalculationResultMock} />
      );
      const operationStringRepresentationActual = wrapper
        .find('.calculation-result-representation')
        .text();
      expect(operationStringRepresentationExepected).toBe(
        operationStringRepresentationActual
      );
    });
  });
});
