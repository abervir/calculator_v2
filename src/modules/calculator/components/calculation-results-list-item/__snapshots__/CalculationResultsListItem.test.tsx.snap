// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculationResultsListItem should render component 1`] = `
<Fragment>
  <td
    className="calculation-result-title"
  >
    Operation1
     operation
  </td>
  <td
    className="calculation-result-representation"
  >
    0 + 1 = 1
  </td>
  <td>
    Thu, 01 Jan 1970 00:00:00 GMT
  </td>
  <td
    className="controls-column"
  >
    <Button
      onClick={[Function]}
      title="Copy result!"
    />
  </td>
</Fragment>
`;
